### ISSUE TITLE
Type inference fails to narrow return type of async helper with union inputs

### ISSUE THREAD

**Issue by vdusek** [astral-sh/ty#2661]
### Description

When using a helper that accepts either a value or a coroutine and always returns the unwrapped value, the type checker cannot infer that the return type is the concrete type `T`.

Even with `overloads`, ty treats a union input as a single type and preserves the union in the return type, leading to false positives when accessing attributes on the result.

### Minimal reproduction

```python
from typing import TypeVar, Coroutine, Any, overload

T = TypeVar("T")

@overload
async def maybe_await(value: Coroutine[Any, Any, T]) -> T: ...
@overload
async def maybe_await(value: T) -> T: ...

async def maybe_await(value: T | Coroutine[Any, Any, T]) -> T:
    if hasattr(value, "__await__"):
        return await value
    return value
```

Usage:

```python
async def get_data() -> int | Coroutine[Any, Any, int]:
    ...

result = await maybe_await(get_data())

reveal_type(result)  # ty infers: int | Coroutine[Any, Any, int]
result.bit_length()  # error: possibly-missing-attribute
```

Expected: result inferred as `int`.

### Version

0.0.14

**Comment by carljm** [comment]
Thanks for the report! This is a slightly more complex case than the one described in https://github.com/astral-sh/ty/issues/2657, but the root cause is exactly the same, so I'll close this as a duplicate. (Quite a coincidence that both were filed on the same day!)
